### ISSUE TITLE
[Epic] Align Vavr 1.0 collections to the completely re-designed Scala 2.13 collections

### ISSUE THREAD

**Issue by danieldietrich** [vavr-io/vavr#2308]
See [Scala 2.13 Collections Rework](https://www.scala-lang.org/blog/2017/02/28/collections-rework.html)

**Comment by nfekete** [comment]
Bite the bullet, go for simulated higher kinded types and unite `*.transform(...)` with `Iterable.to(...)`? Because they are basically the same, except for the lack of HTKs, right?

**Comment by danieldietrich** [comment]
@nfekete I know you don't like a generic `to(...)` conversion method. But explicitly enumerating all possible conversion methods has drawbacks and does not scale well. The solution described above does not simulate HKT's. It is just a compositional approach using functions. It aligns to the Scala 2.13 collection rework. Vavr's `.ofAll` factory method is called `fromIterable` in Scala. I only we could express on the type level that a type `Class<C>` has a specific static method...

**Comment by nfekete** [comment]
I'm not for the enumerating all the conversion methods at all. I think a single `transform(...)` would be enough, if we could only express it in Java without simulated HKTs and other workarounds. @jbgi's work on [derive4j/hkt](https://github.com/derive4j/hkt) seems very interesting. There is also [KindedJ](https://github.com/KindedJ/KindedJ) which seems to be adopted by [Arrow for Kotlin](https://github.com/arrow-kt/arrow) and [AOL's Cyclops](https://github.com/aol/cyclops-react)

But I also recognize your aim to simplify things as much as possible and staying away of external dependencies.

**Comment by danieldietrich** [comment]
Yeah, I know. I was involved when the KindedJ Github org was founded. (In fact I'm an admin of that org.)

However, that's not the right direction for Vavr. HKTs are needed by the projects you mentioned in order to lift functions to a level that operates on Monads. HKTs are a dirty workaround in Java, they have limitations (e.g. only one inheritance level configurable). Such algebraic abstractions are completely out of scope for Vavr. We strive for a dead-simple library for 'Avarage Java Joe' that hides away explicit abstractions like Functor, Monad, Monoid, Applicative, Lense, etc.

**Comment by emmanueltouzery** [comment]
I'm a little curious though, about `to` versus `transform`. am I missing something, or every call to `to` could be replaced by a call to `transform`?

If that's the case, maybe cave could have only the latter, to reduce the API surface area?

**Comment by danieldietrich** [comment]
@emmanueltouzery `to` is shorter and also present in Scala. `transform` will disappear. We align to Scala in the first place (and when in doubt). We could introduce the 0.9.2 methods for backcompat and deprecate them. But I would prefer to remove them.

**Comment by danieldietrich** [comment]
@emmanueltouzery you (and @nfekete) are right! We can widen the codomain of `to`:

```java
default <C> C to(Function<Iterable<T>, C> fromIterable) {
    return fromIterable.apply(this);
}
```

**Comment by nfekete** [comment]
But they are not the same.
Vector:
```
<U> U transform(Function<? super Vector<T>, ? extends U> f)
```
Map:
```
<U> U transform(Function<? super Map<K, V>, ? extends U> f)
```
Iterable:
```
<U, C extends Iterable<U>> C to(Function<Iterable<T>, C> fromIterable)
```

There is value is the specificity of the `transform` method as opposed to the genericity of the `to` method. Transform takes a function that takes as input the concrete type on which transform is defined. Hence, you can efficiently transform a `Vector` by directly indexing one of the values in it, or directly taking out a value from a `Map`. You can't do that with the current version of `to` because it's too generic, it's takes a function that takes an `Iterable` of values. You can't take advantage of the concrete type when doing the transformation.

The two methods _could_ be united, but only if you could somehow declare at the top level (at `Iterable`) that you're gonna take a function that takes as input the current _self_ type. But you cannot express this without HKT's I think.
```
<U> U transform(Function<? super *SELF*, ? extends U> f)
```

Because we can't express the above, I think it's better to keep both. `transform` methods on the concrete implementation level, `to` on `Iterable`.

**Comment by danieldietrich** [comment]
@nfekete That's right. A transformation is just a function. When removing `transform` we loose the fluent API.

In a first step I will take a look at the existing Scala collections API. It is pretty mature. Afterwards I think about what to do additionally...

**Comment by emmanueltouzery** [comment]
Just a data point: I like the fluency of transform and would regret seeing it go. On the other hand, I can see why just removing the to* without adding a to() would leave people confused at how to port these calls.

In the end, if it was me, I think today in vavr, the to* calls and the transform calls serve different functions, and would keep both options.

**Comment by danieldietrich** [comment]
@nfekete, @emmanueltouzery I also think both, to() and transform(), are helpful. We will implement both in Vavr 1.0.

Also I've thought much about modularity. The first module was a test-balloon. I care about you and existing code-bases. The more I think about modularization and the benefits of the current Vavr 0.9.x monolith, the more I think I need to keep the main functionality of Vavr _in one module_.

I will prepare a suggestion that preserves most of the existing functionality.

**Comment by danieldietrich** [comment]
## Type Hierarchy

![io.vavr.collection](https://user-images.githubusercontent.com/743833/46982968-c9b1f400-d0de-11e8-9714-37dc34a69960.png)

We will have the following implementations:

* HashSet and HashMap (probably based on CHAMP as in Scala 2.13, not on HAMT)
* TreeSet and TreeMap
* Queue, List and LazyList (LazyList has also a lazy head and will replace Stream)
* Array, Vector

All other collections will go to an 'extended collections' module

**Comment by danieldietrich** [comment]
## Serializability

* interfaces will not extend Serializable
* collection classes will extends Serializable (and are probably final)

We need to take care of casting Comparators of sorted collections to (Comparator & Serializable).

**Comment by danieldietrich** [comment]
Thank you @tlinkowski, that is really helpful. I tweeted your lovely blog post link: https://twitter.com/danieldietrich/status/1083362477147127810

I will experiment with it!
